### PR TITLE
feat(compete): registration add-ons (merch) gated by entitlement

### PR DESCRIPTION
**Pull request by zacjones93**
# Registration Add-ons (Merch), Gated by Entitlement

Implements **Option A — in-flow registration add-ons** from the [merch add-ons research memo](https://wodsmith-merch-addons-research.wzrrd.sh/): organizers sell merch (e.g., event tees with sizes) inside the registration flow. Selections become extra line items in the same Stripe Checkout Session; pickup is at the venue. Implementation plan: `docs/plans/registration-addons-plan.md`.

## Admin control (the entitlement gate)

Selling add-ons requires the new **`registration_addons`** team feature. Because `/admin/entitlements` lists features straight from the DB, the feature appears in the existing admin UI automatically — platform admins grant/revoke merch selling per organizing team with zero new admin surface. Enforcement is server-side everywhere:
- CRUD server fns throw without the feature
- The athlete-facing catalog returns empty if it's missing (covers revoke-after-create) or the organizer has no verified Stripe account
- `initiateRegistrationPaymentFn` rejects `addOns` input for unentitled teams
- The organizer Merch page renders a locked state

## What's included

- **Schema**: `competition_products` + `competition_product_variants` (sizes, optional per-variant stock); `commerce_purchases` gains `variantId` + `quantity`
- **Checkout**: add-ons ride the existing multi-line-item session; per-unit all-in pricing (percentage platform fee only — no $2 fixed fee on merch) so summary, Stripe line item, and purchase row are cent-identical; coupons never discount merch; free division + paid shirt routes through Stripe
- **Workflow**: ADDON branch completes purchases without creating registrations — transactional variant stock claim with the PENDING→COMPLETED flip as the idempotency gate, auto-refund (`reverse_transfer`) on oversell, group refund when every registration in the session failed capacity re-check
- **Availability**: `availableUntil` order-by deadline (end-of-day in competition timezone, like `registrationClosesAt`) and/or per-variant stock; deadline-only is the recommended default
- **Organizer UI**: Merch page (catalog CRUD, print-shop counts-by-variant report, per-athlete pickup list) + sidebar entry
- **Athlete UI**: "Event merch" order-bump section in both registration form variants, fee summary line items
- **Hardening** (from a 7-angle self-review): addon purchases excluded from competition capacity counts and revenue stats (`isNotNull(divisionId)`), invite settling for free invited divisions routed through Stripe by add-ons, fee summary for free divisions, and a Stripe `application_fee_amount` clamp fixing a latent crash when large coupons meet organizer-absorbed fees
- **Docs**: organizer how-to (`apps/docs/.../sell-merch.md`), athlete tutorial mention, lat.md knowledge sections

## Verification

2,850 tests passing (30 new), type-check + Biome clean, `lat check` green, docs site builds.

## Rollout checklist

- [ ] Apply schema (`pnpm db:push` to the PlanetScale branch, or generate the migration at merge per the local-db workflow — the repo has no MySQL migration journal yet)
- [ ] Insert the `feat_registration_addons` row into `features` in production (mirrors `scripts/seed/seeders/02-billing.ts`)
- [ ] Grant the feature to the first organizer team at `/admin/entitlements`

## Open question (non-blocking)

Platform fees are computed from undiscounted totals (coupons are organizer-funded). The new clamp prevents Stripe rejecting sessions when a large coupon meets organizer-absorbed fees, but whether fees should *scale down* with discounts in general is a fee-policy decision left unchanged.

https://claude.ai/code/session_016sJpPZW1vc5nuMxeCtw9vi

---
_Generated by [Claude Code](https://claude.ai/code/session_016sJpPZW1vc5nuMxeCtw9vi)_

---
## Summary by cubic
Adds registration add-ons (merch) inside the registration flow, gated by the `registration_addons` team entitlement. Athletes can add merch to the same Stripe Checkout; organizers get a simple merch catalog and fulfillment reports.

- **New Features**
  - Entitlement gate: `registration_addons` controls access; enforced in CRUD, public catalog, and checkout.
  - Schema: `competition_products` and `competition_product_variants`; `commerce_purchases` adds `variantId` and unsigned `quantity` (with an index on `variantId`).
  - Athlete UI: “Event merch” in registration with fee summary lines; coupons never discount merch; free division + paid merch still uses Stripe.
  - Organizer UI: Merch page (catalog CRUD, counts-by-variant report, pickup list) with a new sidebar link; shows locked state when not entitled.
  - Checkout: ADDON branch with transactional stock claim and auto-refund on oversell; group refund when all registrations in a session fail.
  - Availability: order-by deadline (end-of-day in competition timezone), per-variant stock, and per-athlete quantity caps.
  - Fees: percentage platform fee only (no fixed $2 on merch) and an application-fee clamp to the post-discount charge to prevent Stripe session rejection.
  - Hardening: ADDON purchases excluded from capacity counts (including public pending sums), per-division revenue stats, and series rollups; `updateCompetitionAddonFn` now updates product + variants in one transaction; merch dialog rejects decimal stock/max-per-athlete; docs and tests added.

- **Migration**
  - Apply schema changes (run your usual DB push/migration).
  - Insert `feat_registration_addons` into `features` (mirrors `apps/wodsmith-start/scripts/seed/seeders/02-billing.ts`).
  - Grant `registration_addons` to the organizer team in `/admin/entitlements`.

<sup>Written for commit dfffec90a7d341b17ee4ed88f69ac3121c2eacce. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/wodsmith/thewodapp/pull/511?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Athletes can now purchase merchandise during registration checkout
  * Organizers can create and manage merch products with variants, pricing, optional inventory limits, and availability deadlines
  * Merch selections are included in registration payments with automated oversold refunds
  * Added fulfillment tracking with sales summaries and per-athlete pickup lists

* **Documentation**
  * Added guides for organizers on managing merch and for athletes on purchasing items during registration

### CODE REVIEW

File: apps/wodsmith-start/src/server-fns/registration-fns.ts [hunk 1]
```diff
@@ -878,6 +1029,91 @@ export const initiateRegistrationPaymentFn = createServerFn({ method: "POST" })
...
+      purchaseIds.push(purchaseId)
+
+      await db.insert(commercePurchaseTable).values({
+        id: purchaseId,
+        userId,
+        productId: addonCommerceProduct.id,
+        status: COMMERCE_PURCHASE_STATUS.PENDING,
+        competitionId: input.competitionId,
+        divisionId: null,
+        variantId: variant?.id ?? null,
```
[the remark this answers is not in the document]
Comment (zacjones93) [reply apps/wodsmith-start/src/server-fns/registration-fns.ts]: Good catch — fixed in 5a0d5d3. `getSeriesRevenueStatsFn` now applies the same `isNotNull(divisionId)` registration-only filter as the competition revenue stats and capacity counts; merch revenue stays on the per-competition Merch page report.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by zacjones93** [comment]
Addressed the automated review feedback in dfffec9:

**Fixed (identified by cubic):**
- `updateCompetitionAddonFn` now runs the product-field update and variant reconciliation in a single transaction, so a variant validation error can no longer leave half-applied product changes (P1).
- Merch dialog rejects decimal input for max-per-athlete and variant stock instead of silently truncating via `parseInt`.
- Added an index on `commerce_purchases.variantId`; `quantity` is now an unsigned int.

**Fixed (identified by CodeRabbit):**
- `getPublicCompetitionDivisionsFn` now excludes null-division (ADDON) pending purchases from the grouped count whose sum feeds public competition-wide capacity — a pending merch purchase could otherwise falsely show "Competition Full" on the register page. Swept all remaining `PENDING` count sites; the rest are division-scoped or intentionally addon-inclusive (the post-checkout readiness poll).

**Declined, with reasons:**
- *Entitlement gate on `archiveCompetitionAddonFn`* (cubic P2): intentional. Archiving only reduces exposure — products from unentitled teams are already invisible to athletes — and gating it would trap a revoked team with ACTIVE products that resurface on re-grant. Create/update remain gated.
- *Per-test `@lat:` comments + seed `@lat` anchor* (CodeRabbit): the repo convention is one `@lat:` ref per spec section, placed at the covering test — adding one per `it()` would create duplicate refs, which `lat check` (currently green) flags.
- *lat.md paragraph-length flags* (CodeRabbit): the ≤250-char rule applies to a section's leading paragraph only; all leading paragraphs comply and `lat check` passes.
- *Add-on group-refund race* (CodeRabbit, Major): known and documented (code comment, lat.md, PR description) — an add-on can complete before a sibling registration fails its capacity re-check. Accepted for v1; a session-wide reconciliation step is the right phase-2 shape.
- *Free divisions finalizing before Stripe session creation in mixed checkouts* (CodeRabbit, Major): pre-existing multi-division behavior this PR extends rather than forks; restructuring it to two-phase finalization is out of scope here.
- *Invite-flip zero-row check / email ordering* (CodeRabbit): matches the established pattern in `updateCompetitionInviteStatus` (workflow), where the PENDING-guarded flip is idempotent best-effort and a zero-row outcome is logged, not failed; allocation is enforced by the authoritative re-check.

---
_Generated by [Claude Code](https://claude.ai/code)_
